### DISCUSSION TITLE
reading a vtk surface mesh file

### DISCUSSION THREAD

**Discussion by ahmad681** [idaholab/moose#23707] [Q&A General] [answered]
Hi, I would like to hear your thoughts on this issue. 
I have an output vtk file generated Unstructured grid that generates that outputs geometry like in the image. 
![image](https://user-images.githubusercontent.com/77033621/224828416-051037fb-eb12-432a-825a-52cd0b26c651.png)
So I am generating the vtk output, and I have a the code in MOOSE that basically calculates elastic energies for cylindrical shape objects where everything inside the cylinder is 1 and outside is zero. 
Now here I am trying to do something different, where I have non-arbitrary objects "cylinders" and at every slice in the z-direction, it looks something like this 
![image](https://user-images.githubusercontent.com/77033621/224828868-88455288-a38f-44a4-a205-a54d394ac2a6.png)

So now everything inside these closed surfaces is 1 and outside is zero. How can I read vtk surface mesh files files in moose, if this is something that was done before? otherwise any thoughts on this? 
actually to be more precise, how do you suggest reading non-arbitrary shape objects from other software output into ones and zeros?
Can Moose read vtk format files as input?

**Comment by GiudGiud** [comment]
I think MOOSE can read vtk, and if it cant there's tons of converters, such as Python's meshio to get it to a format we can read.

**Reply by ahmad681** [reply]
it seems like it does
https://mooseframework.inl.gov/source/mesh/FileMesh.html

**Reply by GiudGiud** [reply]
Use this instead
https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html

**Reply by ahmad681** [reply]
actually, there is one more question, the file that is shown above is surface mesh in vtk format, I don't think moose can read surface mesh and converts to volumetric mesh. 
so I am thinking of converting the vtk to stl, and surface mesh can be converted to volume mesh in msh format via gmsh. 
Unless if moose can actually read surface mesh and convert to volume mesh, that would be fantastic

**Reply by GiudGiud** [reply]
so read the surfaces and mesh what's in between?

MOOSE cannot do that indeed at the moment

**Reply by ahmad681** [reply]
well I would have to do surface mesh to volume mesh outside and then pass in to moose following the link above.

**Reply by ahmad681** [reply]
Okay, made progress here and this is getting interesting
I initialzed block mesh 
and then I imported the mesh that I am generating from gmsh "those nanopillars" basically. 

![image](https://user-images.githubusercontent.com/77033621/225126437-354cfdab-e340-4483-b05b-64befa86ec1e.png)

So I am just displacing them apart for visualization, but in theory the pillars should be embedded in the cube. I am using combinergenerator here by the way. 
[Mesh]
  [./fmg]
    type = FileMeshGenerator 
    file = '../surface_mesh.msh'
  [../]
  [./gmg]
    type = GeneratedMeshGenerator
    dim = 3
    elem_type = TET4
    nx = 50
    ny = 50
    nz = 50
    xmin = -20
    xmax = 20
    ymin = -20
    ymax = 20
    zmin = -60
    zmax = 20
  [../]
  [./cmbn]
    type = CombinerGenerator 
    inputs = 'fmg gmg'
  [../]
[]

So to my understanding and I looked up the source code, it seems like combinergenerator generate disjoint mesh and doesn't them combine them into one, however, this is not what I am seeing here. 
I would really like to have two different pointer nodes, one existing in cube, and one existing in the nanopillars. 

This is to be done for physics, because I am looping through *_current_node that is defined in AuxKernel.h 
const Node * const & _current_node; 

and so my question is how can I know that _current_node exists in filemeshgenerator and generatedmeshgenerator separately, 

Hopefully I am making sense here, :)

**Reply by GiudGiud** [reply]
CombinerGenerator makes no attempt at stitiching, and it makes no attempt at removing one volume from the other as well.

So you have two overlapping disjoint meshes right now, which I imagine is not good for your problem. 

You want the nodes at the boundary of one mesh to match with the nodes at the boundaries in the other mesh right?

**Reply by ahmad681** [reply]
"You want the nodes at the boundary of one mesh to match with the nodes at the boundaries in the other mesh right?" 
yes,Exactly, and then end up with two distinct meshes maybe one for pillars and for surrounding region,

**Reply by GiudGiud** [reply]
why two distinct meshes? Do you want to have two different simulations?
Or do you mean two distinct subdomains?

MOOSE does not have the ability to do that.
You may try Cubit (with the python api), gmsh or Salome which should all be able to do that though it may be difficult for some

**Reply by ahmad681** [reply]
sorry mispoke, two distinct subdomains, and one simulation, but one mesh for subdomain 1, and one for subdomain 2.  It's basically like a sharp-interface problem.  
Oh MOOSE can't do MOOSE can't do two subdomains?

**Reply by GiudGiud** [reply]
MOOSE can do two subdomains.

It cant stitch your two meshes unless they are already conforming (have all the nodes in the right place, ready to be merged)

**Reply by ahmad681** [reply]
I see, that makes sense then. 
I will keep working on it. and keep this thread updated if that's okay.

**Reply by GiudGiud** [reply, marked as answer]
Sounds good!
Please let us know what you end up using, we can tag some users of these meshing tools.
If you think you can code an object in MOOSE to make the mesh (meshgenerators) we could also be interested in the contribution

**Reply by ahmad681** [reply]
Awesome, of course, hopefully we come up with something good and efficient. 
to be clear main, the idea is, have mesh for the nanopillars, and mesh for the matrix and solve equations. and do either sharp interface or diffuse interface if I remesh these two domains into one mesh to solve mismatch stresses. haha, both approaches got difficulties honestly.
